### ISSUE TITLE
[temp.names] Current wording doesn't all for deduction of template parameter when a name prefixed with template is not followed by template argument list <> 

### ISSUE THREAD

**Issue by ranaanoop** [cplusplus/CWG#488]
Full name of submitter: Anoop Rana

Reference (section label): [temp.names]

Issue description: The [current wording](https://eel.is/c++draft/temp.names#6.sentence-1) doesn't seem to allow `c.template func(5)` in the below shown program. We see [implementation divergence](https://godbolt.org/z/5oGYYz1P5) here.

```
template<typename T>
struct C
{
     template<typename U> void func(U)
     {

     }
};
int main()
{
    C<int> c;
    c.template func(5); //this is invalid as per current wording       
} 
```
The current wording in [temp.names](https://eel.is/c++draft/temp.names#6.sentence-1) states:

> A name prefixed by the keyword template shall be followed by a template argument list or refer to a class template or an alias template[.](https://eel.is/c++draft/temp.names#6.sentence-1)

**Comment by frederick-vs-ja** [comment]
I think this is just a bug of MSVC and the standard wording looks fine.

**Comment by ranaanoop** [comment]
> I think this is just a bug of MSVC and the standard wording looks fine.

@frederick-vs-ja There was a similar defect/gap in the standard wording [CWG2608 Allowing ommiting empty template argument list for default arguments](https://github.com/cplusplus/CWG/issues/79). I think here also, there is gap/defect in the wording very similar to that.

**Comment by ranaanoop** [comment]
@frederick-vs-ja I don't think this is an msvc issue because msvc starts [accepting](https://godbolt.org/z/Yrj94haa4) the code if you specify an empty argument list `<>`. That is, this does seem like a wording defect in the standard.

```
template<typename T>
struct C
{
     template<typename U> void func(U)
     {

     }
};
int main()
{
    C<int> c;  
    c.template func(5); //this is invalid as per current wording gcc and clang accepts but msvc rejects     
    //c.template func<>(5); //valid as per current wording. All compilers accepts this including msvc
}
```

**Comment by jensmaurer** [comment]
The reason `template` exists in this syntactic position is to be able to differentiate the meaning of the token `<` between a less-than operator and the start of a _template-argument-list_. If there is no _template-argument-list_, `template` is not needed, and, as you correctly observed, in that case it's actually forbidden.

Please file bug reports against the affected vendors; the standard wording seems clear. This case is different from CWG2608, where syntactic disambiguation isn't in an issue.

In your particular example,  `c` is non-dependent, so `template` isn't needed regardless of the presence of a _template-argument-list_ (because the meaning of all names are known).

If you feel you want to make `template` optional (as opposed to forbidden) for the case above, feel free to write a paper with rationale, directed to EWG.
